src/coreclr/gc/gc.cpp

```diff
@@ -7120,36 +7122,48 @@ void gc_heap::fix_allocation_context (alloc_context* acontext, BOOL for_gc_p,
                  (size_t)acontext,
                  (size_t)acontext->alloc_ptr, (size_t)acontext->alloc_limit));
 
+    if (acontext->alloc_ptr == 0)
+    {
+        return;
+    }
+
     int align_const = get_alignment_constant (TRUE);
 
-    if (((size_t)(alloc_allocated - acontext->alloc_limit) > Align (min_obj_size, align_const)) ||
+#ifdef USE_REGIONS
+    heap_segment* region = find_segment(acontext->alloc_limit, FALSE);
+    bool is_ephemeral_heap_segment = region == ephemeral_heap_segment;
+    uint8_t* allocated = is_ephemeral_heap_segment ? alloc_allocated : heap_segment_allocated (region);
+#else // USE_REGIONS
+    uint8_t* allocated = alloc_allocated;
+#endif // USE_REGIONS
+    if (((size_t)(allocated - acontext->alloc_limit) > Align (min_obj_size, align_const)) ||
         !for_gc_p)
     {
         uint8_t*  point = acontext->alloc_ptr;
-        if (point != 0)
-        {
-            size_t  size = (acontext->alloc_limit - acontext->alloc_ptr);
-            // the allocation area was from the free list
-            // it was shortened by Align (min_obj_size) to make room for
-            // at least the shortest unused object
-            size += Align (min_obj_size, align_const);
-            assert ((size >= Align (min_obj_size)));
+        size_t  size = (acontext->alloc_limit - acontext->alloc_ptr);
+        // the allocation area was from the free list
+        // it was shortened by Align (min_obj_size) to make room for
+        // at least the shortest unused object
+        size += Align (min_obj_size, align_const);
+        assert ((size >= Align (min_obj_size)));
 
-            dprintf(3,("Making unused area [%Ix, %Ix[", (size_t)point,
-                       (size_t)point + size ));
-            make_unused_array (point, size);
+        dprintf(3,("Making unused area [%Ix, %Ix[", (size_t)point,
+                    (size_t)point + size ));
+        make_unused_array (point, size);
 
-            if (for_gc_p)
-            {
-                generation_free_obj_space (generation_of (0)) += size;
-                if (record_ac_p)
-                    alloc_contexts_used ++;
-            }
+        if (for_gc_p)
+        {
+            generation_free_obj_space (generation_of (0)) += size;
+            if (record_ac_p)
+                alloc_contexts_used ++;
         }
     }
     else if (for_gc_p)
     {
-        alloc_allocated = acontext->alloc_ptr;
+#ifdef USE_REGIONS
+        if (is_ephemeral_heap_segment)
+#endif // USE_REGIONS
+            alloc_allocated = acontext->alloc_ptr;
```

Comment:
nit - could we please add a {} around this line because now there's an if statement? 

```
{
    alloc_allocated = acontext->alloc_ptr;
}
```